README
HTTPsmash - a distributed load testing system for web infrastructures
================================================================================
Author: Madeline Everett
License: BSD http://www.opensource.org/licenses/bsd-license.php


High Level Feature List
================================================================================
1. Highly scalable HTTP load generation application for simulating high traffic
2. Allows geographically distributed nodes to simulate global user-base traffic
2. Enhanced job management via inter-process queue system
3. Encrypted node communication via direct SOCKET protocol w/ key exchange
4. Optional randomized query string url generation - simulate dynamic calls
5. Multi-threaded operation on server and client 
6. Config file support for server and client
7. Performance data reporting for url connection timing
8. Configurable options for controling total hit quantity across nodes
   - per-node thread concurrency
   - per-thread connect cycling
   - per-connection delay timing
   - optional randomized connection timing


Topic List
================================================================================
+ Scaling Your Load Test Cluster
+ Hardware and Software Requirements
+ Components for Reference
+ HTTP Smash Server
+ HTTP Smash Client
+ Scaling Your Load Test Cluster
+ Application Support


Hardware and Software Requirements
================================================================================

Software
----------------------------------------
Python >=2.6
  modules: socket, stat, binascii, base64, ConfigParser, commands, threading, 
  os, sys, urllib2, select, random, string, getopt, time, logging, queue, 
  thread, fileinput, zlib

Operating Systems (only 64bit tested, probably fine on 32bit)
----------------------------------------
Ubuntu: Ubuntu 10.10+ 
Redhat flavors: 5.5 -> 6.3, Fedora 14+
UNIX: openBSD 4.8+, FreeBSD 8.1+
Mac OSX: 10.5 -> 10.9
Windows: not tested, if you get it to work then that's great

Hardware Requirements
----------------------------------------
See 'scaling' section below. Dedicated hardware will give the best performance.

Network Requirements
----------------------------------------
Basic TCP/IP required
Multicast not required
Firewall needs to allow access between the clients and server but client nodes do not need to be able to talk to one another
Default Manager Port: port 8080
If you have TCP Offload support on your network cards you should enable it for better performance of network saturation


Components for Reference
================================================================================
1. Server (workload manager and task delegation)
  - configuration file
  - workload definition file
  - encryption key management
  - encryption engine
  - workload queue manager
  - threading manager
  - TCP connection socket engine

2. Client (task execution)
  - configuration file
  - encryption key management
  - encryption engine
  - task queue manager
  - threading manager
  - task request engine - pull style

HTTP Smash Server
================================================================================

Configuration
----------------------------------------
The server is very simple. You configure it via two files.
"server/conf/server.conf" - this contains variables that are documented in the conf file
"server/conf/urlfile" - this file is where you put your URLs to test.

Each line in the "urlfile" is equivallent one job. The options are required to be in order, separated by blank spaces. If you do not separate them properly, or leave out a value for a setting then the server will fail to start and will let you know as such.

When creating your jobs please consider the following equations to determine how much load you are going to be hitting your url with. We operate with the following equations for traffic planning:
  * connections per-second = (client-threads * client-perthread) / (client-perthread * wait)
  * connections total = (client-threads * client-perthread) * repeat

Job Entry Values
----------------------------------------
 clients = number of clients you want running the jobs (dependent on the # of active job nodes reading the queue)
 url = url to test
 wait = time period to wait between each url request
 repeat = quantity of times to repeat requests per thread
 suffix = optional query string or url value to append to url (required if using random query strings)
 rand = random query string length
 rand_enable = enable randomized query string if desired, requires >1 for RAND value
 suffix_enable = enable optional query string, requires "RAND_ENABLE = 1"
 client-threads = number of threads the client should open per job (each thread will open $PERTHREAD connections)
 client-perthread = number of connections to the destination per THREAD

Job Example
----------------------------------------
  clients url wait repeat suffix rand rand_enable suffix_enable client-threads client-perthread
  example:
  2 http://url.com 1 4 /string= 8 0 0 2 4


HTTP Smash Client
================================================================================
The client is simple as well. You configure it via the file "client/conf/client.conf"
- Run it and it will connect to the manager and process jobs from the queue in a FIFO manner. 
- Tail the log to see extra info while running


Scaling Your Load Test Cluster
================================================================================
 - You may need to run multiple nodes (or multiple per-node instances of the client) to process very large workloads in a short period of time. 
 - Connection concurrency for the tests are limited to the number of threads that your server(s) can concurrently support. 
 - A standard server grade processor (Xeon / Opteron) at the time of this writing operates with between 2-6 threads per core with 1-4 socket CPUs being the most common commodity hardware. 
 - The more threads your CPU supports more task execution threads you can run = more concurrent load tests able to be run. 
 - To saturate a 1000Mbit (gigabit) ethernet connection you need very high concurrency of executing threads; generally well over 500 connections per second per node. YMMV. 
 - If you have TCP Offload support on your network cards you should enable it for better performance of network saturation, since the NIC's ASIC will process TCP instead of the CPU
 - Moore's law exists - please apply this to your scaling plans.


Application Support
================================================================================
Look at the debug log for errors or troubleshooting information prior to contacting the author or digging into the code.
Professional support is available on a consulting basis





CHANGELOG
# 1.1.5
- import from private repo

queue/queue.py
#!/usr/bin/env python
###################################################################
## DATE: 2010-11-11
## FILE: queue
## LICENSE: BSD http://www.opensource.org/licenses/bsd-license.php
###################################################################
import os
import time
import urllib2
from Queue import *
from threading import Thread, Lock


def do_work(url):
     start = time.time()
     try:
          urllib2.urlopen(url)                
     except:
          print "FAILED|%s"%(url)
          return
     end = time.time()                
     elapse = str(round((end - start)*1000))+"ms"
     print "TIMING|elapse|%s|%s"%(url,elapse)

def worker():
    while True:
        item = q.get()
        do_work(item)
        q.task_done()

def queue_source():
     global THREADS
     fileIN = open("url_list", "r")
     return fileIN.readlines()     

q = Queue()
for i in range(4):
     t = Thread(target=worker)
     t.setDaemon(True)
     t.start()
    
for url in queue_source():
     url = url.rstrip()
     print "inserting into the queue: %s"%(url)
     q.put(url)

#wait for the queue to be processed before quitting
q.join() 


scripts/genkey.py
#!/usr/bin/env python
import sys
import time
import random
try:
    from Crypto.Hash import SHA256 
    HASH_REPS = 1024 #cycles for creating hashes
except:
    raise error("This program requires the Python 2.6 Crypto package")

'''START encryption functions'''
PRIVATE_HASH = 'keys/generator.hash'
STORED_HASHES = 'keys/stored.hash'

def __saltedhash(string, salt):
    sha256 = SHA256.new()
    sha256.update(string)
    sha256.update(salt)
    for x in xrange(HASH_REPS): 
        sha256.update(sha256.digest())
        if x % 10: sha256.update(salt)
    return sha256

def saltedhash_hex(string, salt):
    """returns the hash in hex format"""
    return __saltedhash(string, salt).hexdigest()

def create_randhash():
    for i in range(256):
        str0 = time.time()
        str0 += random.random()        
        
    for i in range(256):
        str1 = time.time()
        str1 += random.random()

    return saltedhash_hex(str(str0),str(str1))


def store_private(userID):
    '''when adding a new client we need to create a server-side-stored unique hash that is compare to their public-key-hash for authentication'''
    '''our private hash is stored in a server-side txt file called "keys/generator.hash", we use this to generate new hashes and compare incomign hashes'''

    try:
        generator = open(PRIVATE_HASH, 'r').read()
    except:
        logger("Failed to open hash file.",'e')
        sys.exit(1)

    userprivate_hash = saltedhash_hex(userID,generator)
    f = open(STORED_HASHES, 'a+')
    f.write(userprivate_hash+"\n")
    f.close()


print "\nThis creates a public key to be used by each unique client that connects to the managerd process. This public key must be stored on the client system in the same directory as the stranglerc app in a file called 'client_key.pub'. The private key for this public key is automatically stored in the keys/stored_keys.hash file on the managerd server.\n\n"
pubhash = create_randhash()
print "PUBLIC-KEY: %s"%(pubhash)
store_private(pubhash)


scripts/findkey.py
#!/usr/bin/env python
import fileinput
import os
import sys
import time
import random
try:
    from Crypto.Hash import SHA256 
    HASH_REPS = 1024 #cycles for creating hashes
except:
    raise error("This program requires the Python 2.6 Crypto package")

'''START encryption functions'''
PRIVATE_HASH = 'keys/generator.hash'
STORED_HASHES = 'keys/stored.hash'

def __saltedhash(string, salt):
    sha256 = SHA256.new()
    sha256.update(string)
    sha256.update(salt)
    for x in xrange(HASH_REPS): 
        sha256.update(sha256.digest())
        if x % 10: sha256.update(salt)
    return sha256

def saltedhash_hex(string, salt):
    """returns the hash in hex format"""
    return __saltedhash(string, salt).hexdigest()

def compare_hash(userHash):
    '''our private hash is stored in a server-side txt file called "keys/generator.hash", we use this to compare incoming hashes'''
    
    try:
        generator = open(PRIVATE_HASH, 'r').read()
    except:
        logger("Failed to open PRIVATE_HASH file: %s."%(PRIVATE_HASH),'e')
        sys.exit(1)

    gen = str(saltedhash_hex(userHash,generator))
    print "chek gen: %s"%(gen)

    if os.path.exists(STORED_HASHES):
        for line in open(STORED_HASHES):
            if gen in line:
                print "0"                
                return 0
            else:
                print "1"
                return 1

    else:
        logger("Failed to open STORED_HASHES file.",'e')
    
pub = str(raw_input("enter public key: "))
ret = compare_hash(pub)
if ret != 1:
    print "MATCH"


server/tmp/http_server.py
#!/usr/bin/env python
from socket import *
from binascii import *
from base64 import *
from stat import *
from ConfigParser import ConfigParser
from bottle import abort, request, route, run, PasteServer
from pymongo import Connection
import bottle as web
import thread
import logging
import sys
import os
import fileinput
import string
import zlib
import json

try:
    from Crypto.Hash import SHA256 
    HASH_REPS = 1024 #cycles for creating hashes
except:
    raise error("This program requires the Python 2.6 Crypto package")

ver = sys.version.split(' ')[0].split(".")
major=ver[:1]
minor=ver[1:2]
version="%s.%s"%(major[0],minor[0])
if version in ('2.4','2.3','2.2','2.1','2.0'):
    pyver = "old"
else:
    pyver = "new"


########### HTTP ROUTES
connection = Connection('localhost', 27017)
db = connection.foo

'''index'''
@route('/')
@route('/index.html')
def index():
    return "<a href='/'>Nothing to see here.</a>"

'''debug controller to show post/get/cookies'''
@web.route('/do/:cmd') #url is: http://host:port/do/show
def cmd(cmd):
    if cmd=="show":
        yield "<li>cookies : %s</li>"% str(web.request.COOKIES)
        yield "<li>get : %s</li>"%     str(web.request.GET)
        yield "<li>post : %s</li>"%    str(web.request.POST)
    else:
        web.redirect("/")

'''insert a job -- needs to be moved to an admin only / auth only method'''
@route('/documents', method='PUT')
def put_document():

    '''incoming data format as follows please: 
    {
    "_id": "%int",
    "org_name": "%str",
    "job_name": "%str, 
    "job_url": "%str", 
    "max_hits_per_second": "%int",
    "concurrent_clients_max": "%int",
    "total_runtime_seconds": "%int",
    "connection_wait_interval": "%int", 
    "query_string_conditional": "%str", 
    "query_string_conditional_enable": "0/1", 
    "query_string_random_length": "%int", 
    "query_string_random_enable": "0/1",
    "datetime_run": "%DT",
    "datetime_created": "%DT",
    "datetime_updated": "%DT"
    }
    '''

    data = request.body.readline()
    if not data:
        abort(400, 'No data received')

    entity = json.loads(data)
    if not entity.has_key('_id'):
        abort(400, 'No _id specified')

    try:
        db['documents'].save(entity)
    except ValidationError as ve:
        abort(400, str(ve))

'''get most recent job from db'''
@route('/documents/:id', method='GET')
def get_document(id):
    logger("getJob page called. id: %s, key: %s"%(id,key),'i')

    if not key:
        abort(404, 'No authentication key provided.') 
        
    retval = compare_hash(key)
    if retval == 1:
        logger("Failed to correctly compare client hash. Retval: %s"%(retval),'e')
        sys.exit(1)

    logger("Hash check: [OK]",'d')       
    entity = db['documents'].find_one({'_id':id})
    if not entity:
        abort(404, 'No document with id %s' % id)
    return entity

############


def encode(data):
    '''This is basic hexify and base64 encode. We need to switch to m2crypt or keyczar eventually'''
    try:
        code = urlsafe_b64encode(hexlify(data))
        return code
    except:
        raise DataCompressionFail()

def print_help():
    print '''==================================================
HTTPsmash_server :: a distributed load testing app
==================================================
Use the following flags to change default behavior
    
   Option                 Description
   --port=                Port to run on
   --buffer=              Buffer size 
   --maxconn=             Maximum connections
   --help                 Print this message
   '''
def logger(detail,level):
    if(level == "d"):
        log.debug("%s"% (detail))
    elif(level == "i"):
        log.info("%s"% (detail))
    elif(level == "w"):
        log.warn("%s"% (detail))
    elif(level == "e"):
        log.error("%s"% (detail))
    elif(level == "c"):
        log.critical("%s"% (detail))

'''START encryption functions'''
PRIVATE_HASH = 'keys/generator.hash'
STORED_HASHES = 'keys/stored.hash'

def __saltedhash(string, salt):
    sha256 = SHA256.new()
    sha256.update(string)
    sha256.update(salt)
    for x in xrange(HASH_REPS): 
        sha256.update(sha256.digest())
        if x % 10: sha256.update(salt)
    return sha256

def saltedhash_hex(string, salt):
    """returns the hash in hex format"""
    return __saltedhash(string, salt).hexdigest()

def compare_hash(resp):
    '''our private hash is stored in a server-side txt file called "keys/generator.hash", we use this to compare incoming hashes'''
    userHash = string.rstrip(resp)
    logger("Received client hash: %s"%(userHash),'d')
    try:
        generator = open(PRIVATE_HASH, 'r').read()
    except:
        logger("Failed to open PRIVATE_HASH file: %s."%(PRIVATE_HASH),'e')
        sys.exit(1)
        
    gen = str(saltedhash_hex(userHash,generator))
    logger("Generated comparison key: %s"%(gen),'d')
    if os.path.exists(STORED_HASHES):
        for line in open(STORED_HASHES):
            logger("Current line comparison key: %s"%(line),'d')
            if gen in line:
                return 0
            else:
                return 1            
    else:
        logger("Failed to open STORED_HASHES file.",'e')
        return 1
'''END encryption functions'''


def main():
    web.debug(DEV)
    web.run(server=PasteServer, host=HOST, port=PORT)

if __name__ == "__main__":
    cfg_file = "conf/server.conf"

    '''test if cfg_file exists, else exit'''
    sys.stdout.write("Testing %s config access: "% (cfg_file))
    if(os.path.exists(cfg_file)):
        sys.stdout.write("[exists]")
        if(os.access(cfg_file, os.R_OK)):
            sys.stdout.write("[readable]\n")
        else:
            sys.stdout.write("[read-failed][exiting(1)]\n")
            sys.exit(1)
    else:
        sys.stdout.write("[config file does not exist][exiting(1)]\n")
        sys.exit(1)

    '''initilize config settings'''
    config = ConfigParser()
    config.read([cfg_file])
    headertxt = 'server configuration'
    LOG = config.get(headertxt,'LOG')
    HOST = config.get(headertxt,'HOST') 
    PORT = int(config.get(headertxt,'PORT'))
    MAXCON = int(config.get(headertxt,'MAXCON'))
    BUFSIZE = int(config.get(headertxt,'BUFSIZE'))
    DEV = str(config.get(headertxt,'DEV'))
    ADDR = (HOST,PORT)    

    global URL
    global WAIT
    global REPEAT
    global CONDITIONAL
    global RAND
    global RAND_ENABLE
    global CONDITIONAL_ENABLE
    global WORDLIST

    '''start logging'''
    log = logging.getLogger()
    log.setLevel(logging.DEBUG)
    formatter = logging.Formatter("%(asctime)s|%(levelname)s|%(message)s")
    c = logging.StreamHandler(sys.stdout)
    c.setLevel(logging.INFO)
    c.setFormatter(formatter)
    f = logging.FileHandler(LOG)
    f.setLevel(logging.DEBUG)
    f.setFormatter(formatter)    
    log.addHandler(c)
    log.addHandler(f)

    logger("running with the following server settings:",'i')
    logger("LOG: %s"%(LOG),'i')
    logger("HOST: %s"%(HOST),'i')
    logger("PORT: %s"%(PORT),'i')
    logger("MAXCON: %s"%(MAXCON),'i')
    logger("BUFSIZE: %s"%(BUFSIZE),'i')

    try:
        main()

    except (KeyboardInterrupt, SystemExit):
        retval = 1
        logger("INTERRUPT: RETVAL: [%s]"%(retval),'e')
        logger("Closing serv.process",'e')
        conn.close()
        sys.exit(1)


server/tmp/https_server.py
#!/usr/bin/env python
'''
SimpleSecureHTTPServer.py - simple HTTP server supporting SSL.

- replace fpem with the location of your .pem server file.
- the default port is 443.

usage: python SimpleSecureHTTPServer.py
'''
import socket, os
from SocketServer import BaseServer
from BaseHTTPServer import HTTPServer
from SimpleHTTPServer import SimpleHTTPRequestHandler
from OpenSSL import SSL

certfile = "keys/server.pem"

class SecureHTTPServer(HTTPServer):
    def __init__(self, server_address, HandlerClass):
        BaseServer.__init__(self, server_address, HandlerClass)
        ctx = SSL.Context(SSL.SSLv23_METHOD)
        ctx.use_privatekey_file (certfile)
        ctx.use_certificate_file(certfile)
        self.socket = SSL.Connection(ctx, socket.socket(self.address_family, self.socket_type))
        self.server_bind()
        self.server_activate()


class SecureHTTPRequestHandler(SimpleHTTPRequestHandler):
    def setup(self):
        self.connection = self.request
        self.rfile = socket._fileobject(self.request, "rb", self.rbufsize)
        self.wfile = socket._fileobject(self.request, "wb", self.wbufsize)


def test(HandlerClass = SecureHTTPRequestHandler,
         ServerClass = SecureHTTPServer):
    server_address = ('127.0.0.1', 443) # (address, port)
    httpd = ServerClass(server_address, HandlerClass)
    sa = httpd.socket.getsockname()
    print "Serving HTTPS on", sa[0], "port", sa[1], "..."
    httpd.serve_forever()


if __name__ == '__main__':
    test()


client/HTTPsmash_client.py
#!/usr/bin/python
###################################################################
## DATE: 2010-11-11
## FILE: HTTPsmash_client
## LICENSE: BSD http://www.opensource.org/licenses/bsd-license.php
###################################################################
from __future__ import division
from socket import *
from stat import *
from binascii import *
from base64 import *
from ConfigParser import ConfigParser
import commands
import threading
import os
import sys
import urllib2
import select
import random
import string
import getopt
import time
import logging

ver = sys.version.split(' ')[0].split(".")
major=ver[:1]
minor=ver[1:2]
version="%s.%s"%(major[0],minor[0])
if version in ('2.4','2.3','2.2','2.1','2.0'):
    pyver = "old"
else:
    pyver = "new"

def logger(detail,level):
    if(level == "d"):
        log.debug("%s"% (detail))
    elif(level == "i"):
        log.info("%s"% (detail))
    elif(level == "w"):
        log.warn("%s"% (detail))
    elif(level == "e"):
        log.error("%s"% (detail))
    elif(level == "c"):
        log.critical("%s"% (detail))

#for memory tracking - START
_proc_status = '/proc/%d/status' % os.getpid()
_scale = {'kB': 1024.0, 'mB': 1024.0*1024,
          'KB': 1024.0, 'MB': 1024.0*1024}

def _VmB(VmKey):
    #given a VmKey string, returns a number of bytes
    #get psedo file /proc/<pid>/status
    try:
        t = open(_proc_status)
        v = t.read()
        t.close
    except IOError:
        return 0.0 #Non Linux system?
    #get VmKey line eg: 'VmRSS: 9999 kB\n...'
    i = v.index(VmKey)
    v = v[i:].split(None, 3) #SPLIT on #of whitespace
    if len(v) < 3:
        return 0.0 #invalid format
    #convert Vm value to bytes
    return float(v[1]) * _scale[v[2]]

def memory(since=0.0):
    #return virtual memory in bytes
    return _VmB('VmSize:') - since

def resident(since=0.0):
    #return resident memory in bytes
    return _VmB('VmRSS:') - since

def stacksize(since=0.0):
    #return stack size in bytes
    return _VmB('VmStk:') - since
#for memory tracking - END

class threader(threading.Thread):
    def __init__(self):
        threading.Thread.__init__(self)
    def run(self):
        global url
        global PERTHREAD
        global u
        global suffix
        global rand
        for i in range(PERTHREAD):
            if wait > 0:
                w = round(random.random(),2) + wait
            else:
                w = round(random.random(),1)
            time.sleep(w)
            if suffix == 1 and random > 0:
                str = wordgen()
                url = "%s%s%s"%(u,suffix,str)
            else:
                url = u
            
            logger("OPER|wait: %s|url: %s"%(w,url),'i')
            try:
                start = time.time()
                urllib2.urlopen(url)                
                end = time.time()                
                elapse = end - start
                logger("TIMING|elapse: %s"%(elapse),'i')
            except:
                global ERROR
                ERROR = 1
                logger("Failed to connect to url.",'e')

def randstr(length):
    global url
    twoletters = [c+d for c in string.letters for d in string.letters]
    r = random.random
    n = len(twoletters)
    l2 = length//2
    lst = [None] * l2
    for i in xrange(l2):
        lst[i] = twoletters[int(r() * n)]
        if length & 1:
            lst.append(random.choice(string.letters))

    return "".join(lst)

def FileNotFoundError():
    logger("ERROR: 0021 - Specified wordlist file not found or cannot be read.","i")
    sys.exit(1)

def wordgen():
    var = []
    for v in range(96):
        var.append(randstr(rand))
    return random.choice(var).strip().lower()
            
def looper(loop):
    global suffix
    if random > 0:
        suffix = ""

    conn_persecond = (int(THREADS) * int(PERTHREAD)) / (int(PERTHREAD) * int(wait))
    conn_total = (int(THREADS) * int(PERTHREAD)) * int(repeat)

    logger("---------------------------------------------------",'i')
    logger("Running with the following settings:",'i')
    logger("threads to spawn: %i"%(THREADS),'i') 
    logger("sequential connections per thread: %i"%(PERTHREAD),'i')
    logger("url to strangle: %s"%(u),'i')
    logger("avg wait between connections: %i"%(wait),'i')
    logger("total connections to make: %i"%(conn_total),'i')
    logger("connections per second: %i"%(conn_persecond),'i')
    logger("url suffix append string: %s"%(suffix),'i')
    logger("random string generator char length: %i"%(rand),'i')
    logger("---------------------------------------------------",'i')

    for i in range(loop):
        logger("Running thread loop: %i"%(i),'i')
        retval = init_thread()
        if retval == 1:
            return 1
    return 0        

def init_thread():    
    try:
        backgrounds = []
        for thread in range(THREADS):
            logger("Spawning thread: %s"%(thread),'d')
            background = threader()        
            background.start()
            backgrounds.append(background)
        for background in backgrounds:
            background.join()
        return 0
    
    except:
        return 

def checkfile(file):
    sys.stdout.write("Testing %s access: "% (file))
    if(os.path.exists(file)):
        sys.stdout.write("[exists]")
        try:
            os.ftruncate(file, 0)
            sys.stdout.write("[truncated]\n")
            os.chmod(file,S_IRUSR | S_IWUSR)
            return 0
        except:
            cmd = "echo ''>> %s"%(file)
            retcode, output = commands.getstatusoutput(cmd)
            if retcode != 0:
                sys.stdout.write("[truncate-failed][exiting(1)]\n")
                sys.exit(1)
            else:
                sys.stdout.write("[truncated]\n")
                os.chmod(file,S_IRUSR | S_IWUSR)
                return 0
    else:
        sys.stdout.write("[create-attempt]")
        if(open(file,"a+")):
            sys.stdout.write("[writeable]\n")
            os.chmod(file,S_IRUSR | S_IWUSR)
            return 0
        else:
            sys.stdout.write("[write-failed]\n")
            sys.exit(1)
    

def print_help():
    print '''==================================================
HTTPsmash_client :: per-node client script 
==================================================

REQUIRED SETTINGS

   Flag                   Descrition 
   --logfile=             File to log operations (default ./strangler.log)
   --help                 Print this message

   '''


def decode(data):
    '''simple, decode the data we're processing.'''
    code = unhexlify(urlsafe_b64decode(data))
    logger("OPER.DECODE: %s"%(code),'d')
    return code


def commIO(str):
    '''communication input/output with management server'''
    try:
        var = decode(conn.recv(BUFSIZE))
        if len(var) == 0:
            sys.exit(1)
        logger("OPER.COMM - data var[%s]: [OK]"%(str),'i')
        conn.send("0")
    except:
        logger("OPER.COMM - data var[%s]: [FAIL]"%(str),'i')
        sys.exit(1)
        return 127 #we're having a return value issue if this fails... needs to be addressed
    
    return var


class commVars:
    '''We iterate through each item needed to run a job from the queue'''

    def get(self):
        self.items = {"url":"", "wait":"", "repeat":"", "suffix":"", "rand":"", "THREADS":"", "PERTHREAD":""}
        self.url = str(commIO("url"))
        self.wait = float(commIO("wait"))
        self.repeat = int(commIO("repeat"))
        self.suffix = str(commIO("suffix"))
        self.rand = int(commIO("rand"))
        self.THREADS = int(commIO("client_threads"))
        self.PERTHREAD = int(commIO("client_perthread"))
            
def main():    
    global conn
    global ERROR
    ERROR = 0
    timeout = int(5)    
    conn = socket(AF_INET,SOCK_STREAM)
    conn.settimeout(timeout)

    try:
        conn.connect((ADDR))
        logger("OPER.INIT - job manager connection at %s:%i: [CONNECTED]"%(HOST,PORT),'i')
    except:
        errno, errstr = sys.exc_info()[:2]
        logger("OPER.INIT - job manager connection at %s:%i: [FAILED:%s]"%(HOST,PORT,errstr),'i')
        return 1        

    data = conn.recv(BUFSIZE)
    if decode(data) == "EMPTY":
        logger("OPER.INIT - server job queue is empty. No work to do. Quitting.",'i')
        return 2

    if decode(data) == "INIT":
        logger("OPER.INIT - data stream decode operation: [SUCCESS]",'i')
        
        try:
            pubkey = open("keys/client_key.pub", 'r').read() 
        except:
            logger("OPER.READ - failed to open client_key.pub file.",'e')

        conn.send(pubkey) #send our public key for authentication, if we are valid we continue
        global u
        global wait
        global repeat
        global suffix
        global rand
        global THREADS
        global PERTHREAD
        
        job = commVars()
        job.get()
        
        u = job.url
        wait = job.wait
        repeat = job.repeat
        suffix = job.suffix
        rand = job.rand
        THREADS = job.THREADS
        PERTHREAD = job.PERTHREAD
        retval = looper(repeat)
            
        logger("Retval=%s"%(retval),'d')
        logger("ERROR = %i"%(ERROR),'d')
        if retval == 0 and ERROR == 0:
            try:
                conn.send("job-complete")
                logger("Sending results to manager: [COMPLETE]",'i')
                logger("Final state code: [%s]"%(retval),'d')
                retval = 0
                
            except:
                logger("Sending results to manager: [ERROR]",'i')
                retval = 1                
        else:
            conn.send("job-failed")
            logger("Sending results to manager: [COMPLETE]",'i')
            logger("Final state code: [%s]"%(retval),'d')
            retval = 1
    else:
        conn.send('FAILED')
        logger("Data stream decode operation: [ERROR].",'i')
        retval = 1
        
    conn.close()
    return retval

if __name__ == "__main__":    
    cfg_file = "conf/client.conf"

    '''test if cfg_file exists, else exit'''
    sys.stdout.write("Testing %s config access: "% (cfg_file))
    if(os.path.exists(cfg_file)):
        sys.stdout.write("[exists]")
        if(os.access(cfg_file, os.R_OK)):
            sys.stdout.write("[readable]\n")
        else:
            sys.stdout.write("[read-failed][exiting(1)]\n")
            sys.exit(1)
    else:
        sys.stdout.write("[config file does not exist][exiting(1)]\n")
        sys.exit(1)

    '''initilize config settings'''
    #global THREADS #num threads to open
    #global u #URL to be tested
    #global PERTHREAD #per connection url hits

    config = ConfigParser()
    config.read([cfg_file])
    headertxt = 'client configuration'
    LOG = str(config.get(headertxt,'LOG'))
    HOST = str(config.get(headertxt,'HOST'))
    PORT = int(config.get(headertxt,'PORT'))
    BUFSIZE = int(config.get(headertxt,'BUFSIZE'))
    ADDR = (HOST,PORT)

    try:
        options, remainder = getopt.getopt(
            sys.argv[1:], '', ['logfile=',
                               'help'])
    except getopt.GetoptError, err:
        print str(err) 
        sys.exit(2)
        
    for opt, arg in options:
        if opt in ('--logfile'):
            LOG = str(arg)
        elif opt in ('--help'):
            print_help()
            sys.exit(2)

    try:
        LOG
        checkfile(LOG)
    except:
        checkfile(LOG)

    #create log instance
    log = logging.getLogger()
    log.setLevel(logging.DEBUG)
    formatter = logging.Formatter("%(asctime)s|%(levelname)s|%(message)s")
    c = logging.StreamHandler(sys.stdout)
    c.setLevel(logging.INFO)
    c.setFormatter(formatter)
    f = logging.FileHandler(LOG)
    f.setLevel(logging.DEBUG)
    f.setFormatter(formatter)
    log.addHandler(c)
    log.addHandler(f)
    #end log creation

    try:
        retval = 0
        while retval == 0:
            retval = main()
            
    except (KeyboardInterrupt, SystemExit):
        logger("DIED: CODE: [%s]"%(retval),'i')
        sys.exit(1)
    else:
        if retval == 1:
            code = "FAIL"
        elif retval == 2:
            code = "WORK COMPLETE"
        elif retval == 0:
            code = "OK"
        logger("FINAL STATE: [%s]"%(code),'i')
        sys.exit(retval)







